llo/plugin.go

```diff
@@ -0,0 +1,924 @@
+			p.Logger.Warn("Cannot add channel, outcome already contains maximum number of channels",
+				"maxOutcomeChannelDefinitionsLength", MAX_OUTCOME_CHANNEL_DEFINITIONS_LENGTH,
+				"addChannelDefinition", defWithID,
+			)
+			continue
+		}
+		outcome.ChannelDefinitions[defWithID.ChannelID] = defWithID.ChannelDefinition
+	}
+
+	/////////////////////////////////
+	// outcome.ValidAfterSeconds
+	/////////////////////////////////
+
+	// ValidAfterSeconds can be non-nil here if earlier code already
+	// populated ValidAfterSeconds during promotion to production. In this
+	// case, nothing to do.
+	if outcome.ValidAfterSeconds == nil {
+		previousObservationsTimestampSeconds, err := previousOutcome.ObservationsTimestampSeconds()
+		if err != nil {
+			return nil, fmt.Errorf("error getting previous outcome's observations timestamp: %v", err)
+		}
+
+		outcome.ValidAfterSeconds = map[commontypes.ChannelID]uint32{}
+		for channelID, previousValidAfterSeconds := range previousOutcome.ValidAfterSeconds {
+			if err := previousOutcome.IsReportable(channelID); err != nil {
+				p.Logger.Debugw("Channel is not reportable", "channelID", channelID, "err", err)
+				// was reported based on previous outcome
+				outcome.ValidAfterSeconds[channelID] = previousObservationsTimestampSeconds
+			} else {
+				p.Logger.Debugw("Channel is reportable", "channelID", channelID)
+				// TODO: change log level based on what type of error we got
+				// was skipped based on previous outcome
+				outcome.ValidAfterSeconds[channelID] = previousValidAfterSeconds
+			}
+		}
+	}
+
+	observationsTimestampSeconds, err := outcome.ObservationsTimestampSeconds()
+	if err != nil {
+		return nil, fmt.Errorf("error getting outcome's observations timestamp: %w", err)
+	}
+
+	for channelID := range outcome.ChannelDefinitions {
+		if _, ok := outcome.ValidAfterSeconds[channelID]; !ok {
+			// new channel, set validAfterSeconds to observations timestamp
+			outcome.ValidAfterSeconds[channelID] = observationsTimestampSeconds
+		}
+	}
+
+	// One might think that we should simply delete any channel from
+	// ValidAfterSeconds that is not mentioned in the ChannelDefinitions. This
+	// could, however, lead to gaps being created if this protocol instance is
+	// promoted from staging to production while we're still "ramping up" the
+	// full set of channels. We do the "safe" thing (i.e. minimizing occurrence
+	// of gaps) here and only remove channels if there has been an explicit vote
+	// to remove them.
+	for _, channelID := range removedChannelIDs {
+		delete(outcome.ValidAfterSeconds, channelID)
+	}
+
+	/////////////////////////////////
+	// outcome.StreamMedians
+	/////////////////////////////////
+	outcome.StreamMedians = map[commontypes.StreamID]*big.Int{}
+	for streamID, observations := range streamObservations {
+		sort.Slice(observations, func(i, j int) bool { return observations[i].Cmp(observations[j]) < 0 })
+		if len(observations) <= p.F {
+			continue
```

Comment:
I thought we need 2f+1 observations?

Reply:
It's a good question. This probably needs to be refactored into an aggregate function like we do for mercury here: https://github.com/smartcontractkit/chainlink-data-streams/blob/master/mercury/aggregate_functions.go#L23

We have a ticket open for refactoring this entire file and adding more tests: https://smartcontract-it.atlassian.net/browse/MERC-3524

I have pushed a change with a comment that explains this check.

Reply:
Nice - thanks for opening that thread with research clarified the check here